### PR TITLE
Pbtools update

### PR DESCRIPTION
**Pull request by jonasblixt**

### CODE REVIEW

File: tools/pbstate/src/pbstate.c [hunk 1]
```diff
@@ -245,9 +243,11 @@ bool pbstate_is_system_active(pbstate_system_t system)
             return config.enable == PB_STATE_A_ENABLED;
         case PBSTATE_SYSTEM_B:
             return config.enable == PB_STATE_B_ENABLED;
+        case PBSTATE_SYSTEM_NONE:
```
Comment (thebolt) [inline tools/pbstate/src/pbstate.c:212]: This is adding new functionality as part of another unrelated commit (the cmake migraiton), consider splitting in two.

File: tools/pbstate/src/main.c [hunk 2]
```diff
@@ -134,10 +134,15 @@ int main(int argc, char * const argv[])
         }
     }
 
+#if !defined(PRIMARY_PART) && !defined(BACKUP_PART)
```
Comment (thebolt) [inline tools/pbstate/src/main.c:169]: I think this should be !defined || !defined (that is, only use the default path below if both are defined). Possibly expressed as !(defined && defined).

File: tools/pbstate/src/pbstate.c [hunk 3]
```diff
@@ -354,3 +354,19 @@ int pbstate_set_system_verified(pbstate_system_t system)
     }
     return pbstate_commit();
 }
+
+int pbstate_read_board_reg(unsigned int index, uint32_t *value)
+{
+    if (index > (NO_OF_BOARD_REGS - 1))
```
Comment (thebolt) [inline tools/pbstate/src/pbstate.c:337]: To allow for (easier) later extension with more board_regs, I would at least consider numbering them with index == 0 as the _last_ register, and then count backwards.

File: tools/pbstate/src/main.c [hunk 4]
```diff
@@ -116,6 +136,18 @@ int main(int argc, char * const argv[])
             case 'i':
                 flag_show = true;
             break;
+            case 'w':
+                flag_write_board_reg = true;
+                board_reg_index = strtol(optarg, NULL, 0);
```
Comment (thebolt) [inline tools/pbstate/src/main.c:141]: No error handling on bad parsing?

File: tools/pbstate/src/pbstate.c [hunk 5]
```diff
@@ -155,29 +157,30 @@ int pbstate_load(const char *p_device,
...
     fd = open(backup_device, O_RDONLY);
 
-    if (fd == -1) {
-        LOG("Error: Could not open '%s'\n", primary_device);
-        return -1;
-    }
-
-    if (read_config(fd, &config_backup) != 0) {
-        LOG("Could not read backup config data\n");
+    if (fd != -1) {
```
Comment (thebolt) [inline tools/pbstate/src/pbstate.c:173]: As a general style I prefer (and most code seems to be) in a style where the success case is as little indented as possible, that is you try as much as possible to have
```
if (error) {
  .. return or break
}

<success case>
```
If nothing else just to keep consistency, consider changing.

File: tools/pbstate/src/pbstate.c [hunk 6]
```diff
@@ -189,11 +192,10 @@ int pbstate_load(const char *p_device,
...
 
     if (!primary_ok && backup_ok) {
-        LOG("Primary configuration corrupt, using backup\n");
+        LOG("Primary state corrupt, using backup\n");
         memcpy(&config, &config_backup, sizeof(config));
     } else if (!primary_ok) {
-        LOG("Primary configuration corrupt\n");
-        LOG("Backup configuration corrupt\n");
+        LOG("Both primary and backup state corrupt\n");
         err = -1;
```
Comment (thebolt) [inline tools/pbstate/src/pbstate.c:199]: Earlier in the method it returns an -errno, however here it is -1  (which maps ot EPERM, which is probably not what should be returned). Within a single method, either just have 0/-1, or 0/-errno, but don't mix

File: src/boot/ab_state.c [hunk 7]
```diff
@@ -16,26 +16,8 @@
...
-    uint32_t enable;                    /*!< Boot partition enable bits */
-    uint32_t verified;                  /*!< Boot partition verified bits */
-    uint32_t remaining_boot_attempts;   /*!< Rollback boot counter */
-    uint32_t error;                     /*!< Rollback error bits */
-    uint8_t rz[488];                    /*!< Reserved, set to zero */
-    uint32_t crc;                       /*!< State checksum */
-} __attribute__((packed));
-
+// For 'struct pb_boot_state'
+#include "../tools/pbstate/src/pbstate.h"
```
Comment (thebolt) [inline src/boot/ab_state.c:20]: Seems a bit wrong when "core" functionality includes parts from tools.. other way around would feel less incorrect

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -34,8 +34,7 @@ jobs:
     - name: Build pbstate
       run: |
             pushd tools/pbstate
-            autoreconf -fi
-            ./configure
+            cmake .
```
Comment (thebolt) [inline .github/workflows/build.yml:37]: CMake generally don't like at all to build in same directory (which is probably why CI fails, I haven't checked).

**Review by thebolt — CHANGES_REQUESTED** [review]
Functionality wise seems ok, but a few cleanups can be done before merging.

File: tools/pbstate/src/main.c (hunk 2 above)
Comment (jonasblixt) [reply tools/pbstate/src/main.c:169]: fixed

File: tools/pbstate/src/pbstate.c (hunk 1 above)
Comment (jonasblixt) [reply tools/pbstate/src/pbstate.c:212]: fixed

File: src/boot/ab_state.c (hunk 7 above)
Comment (jonasblixt) [reply src/boot/ab_state.c:20]: fixed

File: tools/pbstate/src/pbstate.c (hunk 3 above)
Comment (jonasblixt) [reply tools/pbstate/src/pbstate.c:337]: fixed

File: tools/pbstate/src/pbstate.c (hunk 6 above)
Comment (jonasblixt) [reply tools/pbstate/src/pbstate.c:199]: fixed

File: tools/pbstate/src/main.c (hunk 4 above)
Comment (jonasblixt) [reply tools/pbstate/src/main.c:141]: fixed

File: tools/pbstate/src/pbstate.c (hunk 5 above)
Comment (thebolt) [reply tools/pbstate/src/pbstate.c:174]: Ah, that is my misunderstanding of the logic in the method.

**Review by thebolt — APPROVED** [review]
Looks good to me now
